instrumentation/netty/netty-4-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/common/FutureListenerWrappers.java

```diff
@@ -11,85 +11,83 @@
 import io.netty.util.concurrent.ProgressiveFuture;
 import io.opentelemetry.context.Context;
 import io.opentelemetry.context.Scope;
-import io.opentelemetry.javaagent.instrumentation.api.ContextStore;
+import io.opentelemetry.instrumentation.api.caching.Cache;
 
 public final class FutureListenerWrappers {
+  // Instead of ContextStore use Cache with weak keys and weak values to store link between original
+  // listener and wrapper. ContextStore works fine when wrapper is stored in a field on original
+  // listener, but when listener class is a lambda instead of field it gets stored in a map with
+  // weak keys where original listener is key and wrapper is value. As wrapper has a strong
+  // reference to original listener this causes a memory leak.
```

Comment:
```suggestion
  //
  // also note that it's ok if the value is collected prior to the key, since this cache is only
  // used to remove the wrapped listener from the netty future, and if the value is collected prior
  // to the key, that means it's no longer used (referenced) by the netty future anyways
```